### ISSUE TITLE
addurls from S3 with multiple subdatasets

### ISSUE THREAD

**Issue by TobiasKadelka** [datalad/datalad#3825]
The objective is to turn an S3 bucket into a hierarchy of datasets. The concrete use case is the HCP dataset. We have a complete table of the entire S3 bucket in the format that `addurls` needs. Example:

```
original_url,subject,filename,version
s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json,100206,.xdlm/100206_3T_Diffusion_unproc.json,uh5tsBjch2PHXfhjMv_V.uYo2VkIdFIu
```

The corresponding `addurls` call would be:

```
datalad addurls $2 '{original_url}' '{subject}//{filename}'
```

when executed like this, the command fails with:

```
addurls(error): /data/BnB_USER/Kadelka/next_HCP/HCP/100206/.xdlm/100206_3T_Diffusion_unproc.json (file) [AnnexBatchCommandError: command 'addurl'
Error, annex reported failure for addurl (url='s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json'): {'file': '.xdlm/100206_3T_Diffusion_unproc.json', 'error-messages': [], 'command': 'addurl', 'success': False} [annexrepo.py:add_url_to_file:1857]]
```

This is not informative, as it does not actually communicate the error (maybe `--json-errors` is missing in the annex call. This was done using 0.12rc6).

When the underlying `annex addurl` call is performed by hand, the error is becomes clear:

```
❱ git annex addurl s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json
addurl s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json Configuration does not allow accessing s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json

download failed: Configuration does not allow accessing s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json
failed
git-annex: addurl: 1 failed
```

What is missing is the `datalad` special remote. It would be good, if this would be clear from DataLad's original error message.

Thanks

**Comment by yarikoptic** [comment]
looking at the code path, 3fd47a079895274cdd3ccb61932a0b93d5ce85e7 (which is 0.12.0rc5-202-g3fd47a079) added `--json-error-messages` to all `--json` invocations of annex.  You could confirm that if you run your addurls with heavy debug output (e.g. `-l 2`) and looking at `addurl` invocations.

But unfortunately the problem is deeper in git-annex itself since it doesn't channel error message into json apparently:
```shell
$> git annex addurl --json --json-error-messages s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json                                       
Configuration does not allow accessing s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json
download failed: Configuration does not allow accessing s3://hcp-openaccess/HCP_1200/100206/.xdlm/100206_3T_Diffusion_unproc.json
{"command":"addurl","success":false,"error-messages":[],"file":null}
git-annex: addurl: 1 failed
```
that was with git annex 7.20191009-1~ndall+1  .

Filed an issue with git-annex - https://git-annex.branchable.com/bugs/error_message_in_addurl_is_not_channeled_into_json_record_with_--json-error-messages/

**Comment by kyleam** [comment]
Based on the git-annex report, this should be fixed 890330f0f (make --json-error-messages capture url download errors, 2019-11-12).  That's part of the 7.20191114 release.

**Comment by yarikoptic** [comment]
AFAIK everything should be working with recent enough datalad and annex, so I will close, please reopen if any issue is left to be resolved
